snakemake/ioutils/rule_items_proxy.py

```diff
@@ -0,0 +1,75 @@
+from typing import Union
+from enum import Enum
+
+
+def rule_item_factory(name: str):
+    """Allows to access input, output etc. from statements inside
+    a rule but outside of run/shell etc. blocks. Returns an object that
+    defers evaluation to the DAG phase.
+    """
+    if name == "threads":
+
+        def inner(_wildcards, threads):
+            return threads
+
+        return inner
+    return RuleItemProxy(name)
+
+
+class KeyKind(Enum):
+    ATTRIBUTE = 0
+    ITEM = 1
+
+
+class RuleItemProxy:
```

Comment:
TODO: add a way to also obtain prefixes from such items, e.g. for passing a prefix path.